python/python/bystro/supervised_ppca/_base.py

```diff
@@ -0,0 +1,516 @@
+
+        Returns
+        -------
+        self : object
+            The model
+        """
+
+    @abstractmethod
+    def get_covariance(self):
+        """
+        Gets the covariance matrix defined by the model parameters
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        covariance : np.array-like(p,p)
+            The covariance matrix
+        """
+
+    @abstractmethod
+    def _save_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of tensorflow variables saved
+        """
+
+    def get_precision(self):
+        """
+        Gets the precision matrix defined as the inverse of the covariance
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        precision : np.array-like(p,p)
+            The inverse of the covariance matrix
+        """
+        covariance = self.get_covariance()
+        precision = la.inv(covariance)
+        return precision
+
+    def get_stable_rank(self):
+        """
+        Returns the stable rank defined as
+        ||A||_F^2/||A||^2
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        srank : float
+            The stable rank. See Vershynin High dimensional probability for 
+            discussion, but this is a statistically stable approximation to rank
+        """
+        covariance = self.get_covariance()
+        srank = _get_stable_rank(covariance)
+        return srank
+
+    def transform(self, X):
+        """
+        This returns the latent variable estimates given X
+
+        Parameters
+        ----------
+        X : np array-like,(N_samples,\sum idxs)
+            The data to transform
+
+        Returns
+        -------
+        S : np.array-like,(N_samples,n_components)
+            The factor estimates
+        """
+        prec = self.get_precision()
+        coefs = np.dot(self.W_, prec)
+        S = np.dot(X, coefs.T)
+        return S
+
+    def conditional_score(self, X, idxs, weights=None, sherman_woodbury=False):
+        """
+        Returns the predictive log-likelihood of a subset of data.
+
+        mean(log p(X[idx==1]|X[idx==0],covariance))
+
+        Parameters
+        ----------
+        X : np.array-like,(N,sum(idxs))
+            The data
+
+        idxs: np.array-like,(sum(p),)
+            The observation locations
+
+        weights : np.array-like,(N,),default=None
+            The optional weights on the samples
+
+        Returns
+        -------
+        avg_score : float
+            Average log likelihood
+        """
+        if sherman_woodbury is False:
+            covariance = self.get_covariance()
+            avg_score = _conditional_score(covariance, X, idxs, weights=weights)
+        else:
+            raise NotImplementedError("Sherman woodbury not imlemented yet")
```

Comment:
I don't understand

Reply:
sorry, that was unclear on my part.  I'm suggesting a change to the error message so that the line reads:

```
raise NotImplementedError("Sherman-Woodbury matrix inversion not implemented yet")
```

(or at least we should fix the typo imlemented -> implemented)